civiform/civiform
# Fix AWS policy condition error on file upload question when SAVE_ON_ALL_ACTIONS flag is on

### Description

When the `SAVE_ON_ALL_ACTIONS` flag is on, we want to make sure we save user data whenever the user clicks any button -- "Save&next", "Review", or "Previous". PR https://github.com/civiform/civiform/pull/6651 updated the file upload question to save the file on all button clicks, and it did this by using JavaScript to change the `<input name="success_action_redirect">` part of the file upload form to redirect to different places depending on which button was clicked:
 - "Review" button: https://staging-aws.civiform.dev/programs/92148/blocks/7/updateFile/true/REVIEW_PAGE
 - "Previous" button: https://staging-aws.civiform.dev/programs/92148/blocks/7/updateFile/true/PREVIOUS_BLOCK
 - "Save&next" button: https://staging-aws.civiform.dev/programs/92148/blocks/7/updateFile/true/NEXT_BLOCK

This all worked on a local server, which uses LocalStack as a fake AWS backend. However, it's failing on staging environments, which use the real AWS backend. The error AWS sends us is a policy condition failure:

```
<Error>
<Code>AccessDenied</Code>
<Message>Invalid according to Policy: Policy Condition failed: ["eq", "$success_action_redirect", "https://staging-aws.civiform.dev/programs/92148/blocks/7/updateFile/true/NEXT_BLOCK"]</Message>
<RequestId>YN2BF2JC815X9K46</RequestId>
<HostId>9T43He5YU7Yc6ttnf+sHrP+tLvTLF/Gn3gDSY1bRsMaJVBh7uOo/ohdJbNQwreSIw37RSttLspo=</HostId>
</Error>
```

**Here's what's happening:** The file upload form has a lot of hidden input fields that AWS uses to verify that a file upload request is legitimate. The two input fields we care about for this are:
 - `<input name="success_action_redirect">`, which contains the URL that AWS should send the user to once the file upload succeeds
 - `<input name="Policy">`, which is a base-64 encoded string of policy requirements that AWS should enforce like security tokens, maximum content size, etc. One of the policy requirements is that [the `success_action_redirect` URL should exactly match the redirect URL we specify](https://github.com/civiform/civiform/blob/8314f7442c9e93ed60bb3260f2a37ed1ab71417b/server/app/services/cloud/aws/SignedS3UploadRequest.java#L246-L256).
 
If the `success_action_redirect` URL embedded inside the `<input name="Policy">` element doesn't match the `success_action_redirect` URL listed in the  `<input name="success_action_redirect">` element, then AWS throws this policy condition error. This is intentional and for security reasons: A malicious actor could manually edit the  `<input name="success_action_redirect">` element to be a URL that they own, and then they would be able to see the file names that are being uploaded to our AWS bucket.

But, now that we need to redirect the user to different places depending on what button they clicked, we can't have the policy require an exact match.

**Here's how this PR fixes it:** Luckily, AWS lets us be a bit flexible with the `success_action_redirect` URL in the Policy: We can say that the `success_action_redirect` URL just has to start with a certain prefix. **This PR updates the Policy to make the `success_action_redirect` URL be prefix-matching instead of exact-matching for the file upload question.** We set the prefix to be `https://staging-aws.civiform.dev/programs/92148/blocks/7/updateFile/true`, and that allows the JavaScript to append `/NEXT_BLOCK`, `/PREVIOUS_BLOCK`, or `/REVIEW_PAGE` to the end of the `<input name="success_action_redirect">` element without causing an AWS policy condition failure.

 
## Release notes

### Checklist

#### General

Read the full guidelines for PRs [here](https://github.com/civiform/civiform/wiki/Technical-contribution-guide#creating-a-pull-request)


#### New Features


### Instructions for manual testing

I deployed this change to my custom deploy at https://caitlinshk-dev.civiform.dev and tested these cases:

1. To test with the flag on:
   -  Enable the SAVE_ON_ALL_ACTIONS flag.
   - As an applicant, apply to a program with a file upload question.
   - Upload a file then click "Review".
   - Verify you don't see the error shown in https://github.com/civiform/civiform/issues/6737.
   - Verify you're taken to the review page and that your file was successfully saved.
   - Do the same test for the "Previous" and "Review" buttons.

2. To test "Save&next" with the flag off:
   -  Disable the SAVE_ON_ALL_ACTIONS flag.
   - As an applicant, apply to a program with a file upload question.
   - Upload a file then click "Save&next".
   - Verify you don't see the error shown in https://github.com/civiform/civiform/issues/6737.
   - Verify you're taken to the next page that your file was successfully saved.
   
3. To test "Previous"/"Review" with the flag off:
   -  Disable the SAVE_ON_ALL_ACTIONS flag.
   - As an applicant, apply to a program with a file upload question.
   - Upload a file then click "Previous".
   - Verify you don't see the error shown in https://github.com/civiform/civiform/issues/6737.
   - Verify you're taken to the right page.
   - Verify the file you uploaded **wasn't** saved (since that's the old behavior).
   - Do the same test for the "Review" button.

I also ran `BASE_URL=https://caitlinshk-dev.civiform.dev bin/run-browser-tests-local file.test.ts`, which runs the previously-failing browser tests on my custom deploy, and all tests passed.

### Issue(s) this completes

Fixes #6737
